src/loaders/workspace_downloader/workspace_downloader.py

```diff
@@ -0,0 +1,356 @@
+    )
+
+    required = parser.add_argument_group("required named arguments")
+    optional = parser.add_argument_group("optional arguments")
+
+    # Required flag argument
+    required.add_argument(
+        "--workspace_id",
+        required=True,
+        type=int,
+        help="Workspace addressed by the permanent ID",
+    )
+
+    # Optional argument
+    optional.add_argument(
+        "--root_dir",
+        type=str,
+        default=loader_common_names.ROOT_DIR,
+        help="Root directory.",
+    )
+    optional.add_argument(
+        "--kb_base_url",
+        type=str,
+        default=loader_common_names.KB_BASE_URL,
+        help="KBase base URL, defaulting to prod",
+    )
+    optional.add_argument(
+        "--workers",
+        type=int,
+        default=5,
+        choices=range(1, cpu_count() + 1),
+        help="Number of workers for multiprocessing",
+    )
+    optional.add_argument(
+        "--token_filename",
+        type=str,
+        help="Filename in home directory that stores token",
+    )
+    optional.add_argument(
+        "--delete_job_dir",
+        action="store_true",
+        help="Delete job directory",
+    )
+
+    args = parser.parse_args()
+
+    (workspace_id, root_dir, kb_base_url, workers, token_filename, delete_job_dir) = (
+        args.workspace_id,
+        args.root_dir,
+        args.kb_base_url,
+        args.workers,
+        args.token_filename,
+        args.delete_job_dir,
+    )
+
+    uid = os.getuid()
+    username = os.getlogin()
+
+    job_dir = _make_job_dir(root_dir, loader_common_names.JOB_DIR, username)
+    output_dir = _make_output_dir(
+        root_dir, loader_common_names.SOURCE_DATA_DIR, SOURCE, workspace_id
+    )
+    boolean_flag = False
+    try:
+        # start podman service
+        proc = loader_helper.start_podman_service(uid)
+        boolean_flag = True
+    except:
+        raise Exception("Podman service fails to start")
+    else:
+        # set up conf and start callback server
+        conf = Conf(job_dir, output_dir, workers, kb_base_url, token_filename)
+
+        visited = set(
+            [
+                upa
+                for upa in os.listdir(output_dir)
+                if loader_helper.is_upa_info_complete(output_dir, upa)
+            ]
+        )
+        print("Skipping: ", visited)
+
+        for obj_info in list_objects(workspace_id, conf, FILTER_OBJECTS_NAME_BY):
+            upa = "{6}_{0}_{4}".format(*obj_info)
+            if upa in visited:
+                continue
+            # remove legacy upa_dir to avoid FileExistsError in hard link
+            dirpath = os.path.join(output_dir, upa)
+            if os.path.exists(dirpath) and os.path.isdir(dirpath):
+                shutil.rmtree(dirpath)
+            conf.queue.put([upa, obj_info])
+
+        for i in range(workers + 1):
+            conf.queue.put(None)
+
+        conf.pools.close()
+        conf.pools.join()
+        conf.stop_callback_server()
```

Comment:
This should also go in the `finally` block

Reply:
conf is initiated in `else` block. If in `try` block podman did not start successfully, it will skip `else` and end up in `finally`. It may not recognize what conf is?

Reply:
Resolved for now. Please feel free to reopen.

Reply:
You'd declare conf prior to the try:
```
conf = None
```
and then in the finally
```
if conf:
    conf.stop_callback_server()
```

Reply:
:+1: